### ISSUE TITLE
Sparsify recurrence_matrix and friends

### ISSUE THREAD

**Issue by bmcfee** [bmcfee/librosa#312]
The functions in the `segment` module use dense (ndarray) representations of recurrence matrices.  This obviously doesn't scale for high sample rates and/or long tracks.

We should transition these toward inherently sparse implementations, with optional dense conversion.

**Comment by stefan-balke** [comment]
What do you have in mind?

With scipy.sparse?

Conversion via .todense()?

**Comment by bmcfee** [comment]
Yeah, exactly.

The underlying implementation would have to change substantially, since it's currently based on sorting the distance matrix.

Since we already have sklearn as a dependency (for agglomerative clustering and decomposition methods), it probably makes sense to just punt down to [NearestNeighbors](http://scikit-learn.org/stable/modules/generated/sklearn.neighbors.NearestNeighbors.html) and repack the output in a sparse matrix format  (or just call `kneighbors_graph`).

The tricky parts here will be making the recurrence<->lag functions operate cleanly with sparse inputs.

**Comment by bmcfee** [comment]
... the other tricky part is preserving diagonal suppression.  I'm actually not sure how to do this without changing the semantics.

The current implementation does the following: for each `i`, find `k` nearest neighbors `j` such that `|i-j| >= w`.

Two ways we could make this work with sklearn:

- find `k+2w-1` neighbors, then prune out any within `w` of `i`, and take the `k` smallest remaining.
- implement a custom distance function that returns `np.inf` when points are too close in time.

Both of these are potentially slow, though I think I prefer the first one.

**Comment by bmcfee** [comment]
Upon further thought, option 2 above won't work because it breaks metric properties (and therefore, efficient search data structures), so we're stuck with option 1.

Using `mode='distance'` in `kneighbors_graph` should make it relatively easy to prune out the extraneous links, since the distance is readily available.
